### PR TITLE
HW4: Ali Ammar

### PR DESCRIPTION
**Pull request by ammarali32**

### CODE REVIEW

File: src/main/java/queue/impl/ArrayQueueimpl.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.Queue;
+
+public class ArrayQueueimpl implements Queue<Integer> {
```
Comment (Walingar) [inline src/main/java/queue/impl/ArrayQueueimpl.java]: codestyle

File: src/main/java/queue/impl/Node.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
+package queue.impl;
+
+public class Node {
```
Comment (Walingar) [inline src/main/java/queue/impl/Node.java]: I guess it is a part of LinkedQueueimpl realization

File: src/main/java/queue/impl/Node.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    public Node getPrevious() {
+        return this.previous;
+    }
+
+    public Object getValue() {
+        return this.value;
+    }
+
+    public void addNext(Node x) {
```
Comment (Walingar) [inline src/main/java/queue/impl/Node.java]: add -> set

File: src/main/java/queue/impl/Node.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
+package queue.impl;
+
+public class Node {
+    private Object value;
```
Comment (Walingar) [inline src/main/java/queue/impl/Node.java]: final

File: src/main/java/queue/impl/LinkedQueueimpl.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+    @NotNull
+    @Override
+    public <T> T[] toArray(@NotNull T[] a) {
+        return null;
+    }
+
+    @Override
+    public boolean add(Integer integer) {
+        Node NEW = new Node(integer, null);
```
Comment (Walingar) [inline src/main/java/queue/impl/LinkedQueueimpl.java]: codestyle

File: src/main/java/queue/impl/LinkedQueueimpl.java [hunk 6]
```diff
@@ -0,0 +1,199 @@
+package queue.impl;
+
+
+import kotlin.jvm.internal.Ref;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.*;
+
+public class LinkedQueueimpl implements Queue<Integer> {
```
Comment (Walingar) [inline src/main/java/queue/impl/LinkedQueueimpl.java]: You have a lot of common parts with ArrayQueueimpl, please, extract them to the base class

**Review by Walingar — CHANGES_REQUESTED** [review]

File: src/main/java/queue/impl/ArrayQueueImpl.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+import org.jetbrains.annotations.NotNull;
+
+import java.util.*;
+
+public class ArrayQueueImpl extends BaseQueue<Integer> {
+    private Integer[] elements = new Integer[2];
+
+    @NotNull
+    @Override
+    public Iterator<Integer> iterator() {
```
Comment (Walingar) [inline src/main/java/queue/impl/ArrayQueueImpl.java:32]: It's better to write nested class

File: src/main/java/queue/impl/ArrayQueueImpl.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+            }
+        };
+    }
+
+    private void ensureCapacity(int capacity) {
+        if (capacity <= elements.length) {
+            return;
+        }
+        Integer[] newElements = new Integer[2 * capacity];
+        if (size >= 0) System.arraycopy(elements, 0, newElements, 0, size);
```
Comment (Walingar) [inline src/main/java/queue/impl/ArrayQueueImpl.java]: Brackets

File: src/main/java/queue/impl/LinkedQueueImpl.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
+package queue.impl;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.util.*;
```
Comment (Walingar) [inline src/main/java/queue/impl/LinkedQueueImpl.java:5]: There Iterator is needed only

File: src/main/java/queue/impl/BaseQueue.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+            throw new NoSuchElementException();
+    }
+
+    public void clear() {
+        while (poll() != null)
+            ;
+    }
+
+    public boolean addAll(Collection c) {
+        if (c == null)
```
Comment (Walingar) [inline src/main/java/queue/impl/BaseQueue.java]: Brackets

File: src/main/java/queue/impl/BaseQueue.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        else
+            throw new NoSuchElementException();
+    }
+
+    public void clear() {
+        while (poll() != null)
+            ;
+    }
+
+    public boolean addAll(Collection c) {
```
Comment (Walingar) [inline src/main/java/queue/impl/BaseQueue.java]: Override modifier

File: src/main/java/queue/impl/BaseQueue.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
+package queue.impl;
+
+import java.util.*;
+
+public abstract class BaseQueue<Integer> extends AbstractCollection<Integer> implements Queue<Integer> {
```
Comment (Walingar) [inline src/main/java/queue/impl/BaseQueue.java]: You can use AbstractQueue here

File: src/main/java/queue/impl/BaseQueue.java [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+    protected BaseQueue() {
+        size = 0;
+    }
+
+    public int size() {
+        return this.size;
+    }
+
+    public boolean isEmpty() {
```
Comment (Walingar) [inline src/main/java/queue/impl/BaseQueue.java]: It is already implemented in base class

File: src/main/java/queue/impl/BaseQueue.java (hunk 11 above)
Comment (Walingar) [inline src/main/java/queue/impl/BaseQueue.java]: It is already implemented in the base class

File: src/main/java/queue/impl/BaseQueue.java [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+
+    public Integer element() {
+        Integer x = peek();
+        if (x != null)
+            return x;
+        else
+            throw new NoSuchElementException();
+    }
+
+    public void clear() {
```
Comment (Walingar) [inline src/main/java/queue/impl/BaseQueue.java]: It is already implemented in the base class

File: src/main/java/queue/impl/BaseQueue.java [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+
+    public Integer remove() {
+        Integer x = poll();
+        if (x != null)
+            return x;
+        else
+            throw new NoSuchElementException();
+    }
+
+    public Integer element() {
```
Comment (Walingar) [inline src/main/java/queue/impl/BaseQueue.java]: Override modifier

File: src/main/java/queue/impl/BaseQueue.java [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+        Integer x = peek();
+        if (x != null)
```
Comment (Walingar) [inline src/main/java/queue/impl/BaseQueue.java]: Brackets

File: src/main/java/queue/impl/ArrayQueueImpl.java [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+    public boolean offer(Integer integer) {
+        if (size < elements.length) {
+            add(integer);
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public Integer poll() {
```
Comment (Walingar) [inline src/main/java/queue/impl/ArrayQueueImpl.java:58]: This method should take constant time, but it takes linear instead

**Review by Walingar — CHANGES_REQUESTED** [review]

File: src/main/java/queue/impl/ArrayQueueImpl.java [hunk 18]
```diff
@@ -0,0 +1,98 @@
+package queue.impl;
+
+
+import java.lang.reflect.Field;
```
Comment (Walingar) [inline src/main/java/queue/impl/ArrayQueueImpl.java]: redundant import

File: src/main/java/queue/impl/BaseQueue.java [hunk 19]
```diff
@@ -0,0 +1,15 @@
+package queue.impl;
+
+import java.util.*;
+
+public abstract class BaseQueue<Integer> extends AbstractQueue<Integer> {
```
Comment (Walingar) [inline src/main/java/queue/impl/BaseQueue.java]: You create a generic parameter with name Integer it is not an Integer type here.
Use `public abstract class BaseQueue extends AbstractQueue<Integer>`

File: src/main/java/queue/impl/BaseQueue.java [hunk 20]
```diff
@@ -0,0 +1,15 @@
...
+    int size;
+
+    protected BaseQueue() {
+        size = 0;
+    }
+
+    public int size() {
+        return this.size;
```
Comment (Walingar) [inline src/main/java/queue/impl/BaseQueue.java]: redundant this

File: src/main/java/queue/impl/ArrayQueueImpl.java [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+import java.util.*;
+
+public class ArrayQueueImpl extends BaseQueue<Integer> {
+    private Integer[] elements = new Integer[10];
+    private int head = 0;
+    private int Cap = 10;
+    private int IndexCircularArray = 0;
```
Comment (Walingar) [inline src/main/java/queue/impl/ArrayQueueImpl.java]: codestyle

File: src/main/java/queue/impl/ArrayQueueImpl.java [hunk 22]
```diff
@@ -0,0 +1,98 @@
+package queue.impl;
+
+
+import java.lang.reflect.Field;
+import java.util.*;
+
+public class ArrayQueueImpl extends BaseQueue<Integer> {
+    private Integer[] elements = new Integer[10];
+    private int head = 0;
+    private int Cap = 10;
```
Comment (Walingar) [inline src/main/java/queue/impl/ArrayQueueImpl.java]: codestyle

File: src/main/java/queue/impl/ArrayQueueImpl.java [hunk 23]
```diff
@@ -0,0 +1,98 @@
+package queue.impl;
+
+
+import java.lang.reflect.Field;
+import java.util.*;
+
+public class ArrayQueueImpl extends BaseQueue<Integer> {
+    private Integer[] elements = new Integer[10];
```
Comment (Walingar) [inline src/main/java/queue/impl/ArrayQueueImpl.java]: If you want to change initial capacity you will need to change two places...

File: src/main/java/queue/impl/ArrayQueueImpl.java [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+    @Override
+    public Integer poll() {
+        if (size == 0) {
+            return null;
+        }
+        size--;
+        int value = elements[head];
+        elements[head] = null;
+        head = (head + 1) % Cap;
+        if (size == Cap / 2 && size >= 10) {
```
Comment (Walingar) [inline src/main/java/queue/impl/ArrayQueueImpl.java]: Why do you check size here but for extending you check it inside ensureCapacity? It is strange

Comment (Walingar) [inline src/main/java/queue/impl/ArrayQueueImpl.java]: And one more thing: If you are on the border of capacity changing then you will always extend/shrink size for each add/poll

Comment (Walingar) [inline src/main/java/queue/impl/ArrayQueueImpl.java]: What if I ask you to increase size in 3 times? You will change each `2` literal?

File: src/main/java/queue/impl/LinkedQueueImpl.java [hunk 25]
```diff
@@ -0,0 +1,100 @@
+package queue.impl;
+
+
+import java.util.*;
+
+public class LinkedQueueImpl extends BaseQueue<Integer> {
+    private Node Head, last;
```
Comment (Walingar) [inline src/main/java/queue/impl/LinkedQueueImpl.java]: cotestyle

File: src/main/java/queue/impl/LinkedQueueImpl.java [hunk 26]
```diff
@@ -0,0 +1,100 @@
...
+
+
+    private class LinkedQueueIterator implements Iterator<Integer> {
```
Comment (Walingar) [inline src/main/java/queue/impl/LinkedQueueImpl.java]: code-style: nested class should be after methods

File: src/main/java/queue/impl/LinkedQueueImpl.java [hunk 27]
```diff
@@ -0,0 +1,100 @@
...
+    @Override
+    public Integer peek() {
+        if (size == 0) {
+            return null;
+        }
+        return (int) Head.getValue();
+    }
+
+    public static class Node {
+        private final Object value;
```
Comment (Walingar) [inline src/main/java/queue/impl/LinkedQueueImpl.java]: Why Object? Integer may be?

**Review by Walingar — CHANGES_REQUESTED** [review]

**Comment by Walingar** [comment]
this PR couldn't be merged, please, fix it

File: src/main/java/queue/impl/ArrayQueueImpl.java [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+import java.util.*;
+
+public class ArrayQueueImpl extends BaseQueue {
+    private int head = 0;
+    private final int limit = 10;
+    private int cap = limit;
+    private int indexCircularArray = 0;
+    private Integer[] elements = new Integer[cap];
+
+    private void ensureCapacity(int capacity) {
```
Comment (Walingar) [inline src/main/java/queue/impl/ArrayQueueImpl.java]: this method is hard to understand, please, rewrite it

**Review by Walingar — CHANGES_REQUESTED** [review]

**Review by Walingar — CHANGES_REQUESTED** [review]
code-style issues

**Comment by Walingar** [comment]
And please, rewrite history. Now it contains conflicts with your branch in main repo.

**Comment by ammarali32** [comment]
> And please, rewrite history. Now it contains conflicts with your branch in main repo.

What are the conflicts for me it is written that there are no conflicts and able to merge

**Comment by ammarali32** [comment]
> code-style issues

I just found two blank lines are there any other issues?

**Review by Walingar — APPROVED** [review]

**Pull request merged**